### PR TITLE
Enso language support with parser in VSCode, IGV, etc.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Outline view and completions for Enso code in VSCode. 

### Important Notes

This PR provides the necessary infrastructure for building VSCode extension that includes `enso_parser` library compiled for all supported platforms. 

VSCode extension can now use libraries from `sbt` that are `publishM2`-ready. To make that possible a documentation must have been provided and fixed for those modules - hence so many changes in `.scala` classes.

image

Last, but not least. The outline view and completions display something.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] Manual testing of [enso4vscode-1.36.103.vsix](https://github.com/enso-org/enso/actions/runs/9499946349) performed on Linux, Mac Intel, Mac Arm, and Windows

### CODE REVIEW

**Comment by Akirathan** [comment]
So if I understand that correctly, you are planning to implement the (Netbeans) LSP support via directly reading the IR? And thus skipping our Language server? Using our LS directly, we would need to fetch the suggestion database and somehow provide the completions from there.

**Comment by JaroslavTulach** [comment]
> implement the ... LSP support via directly reading the IR? 

To implement *any* LSP support one needs to implement the processing of [VSCode LSP](https://microsoft.github.io/language-server-protocol/). That requires parsing the individual messages, processing them, computing reply and sending it back.

> ... thus skipping our Language server? 

In spite of the name [Enso Language Server (protocol)](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md) and [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) don't have anything in common.

**Comment by Akirathan** [comment]
> To implement any LSP support one needs to implement the processing of [VSCode LSP](https://microsoft.github.io/language-server-protocol/). That requires parsing the individual messages, processing them, computing reply and sending it back

Right.

> In spite of the name [Enso Language Server (protocol)](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md) and [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) don't have anything in common.

I know, I just wanted to suggest a different approach for LSP implementation - a facade for Enso LS. Some LSP messages are already implemented by Enso LS, e.g. `text/applyEdit`, `text/openFile`, etc. Many are not implemented, but can be inferred from the suggestion database. I was thinking of providing a separate server that would work as a facade for Enso LS, which would forward some messages, receive suggestion database updates from the Enso LS, and process some messages from LSP client by traversing the database.

**Comment by JaroslavTulach** [comment]
> > In spite of the name [Enso Language Server (protocol)](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md) and [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) don't have anything in common.
> 
> I know, I just wanted to suggest a different approach for LSP implementation - a facade for Enso LS. Some LSP messages are already implemented by Enso LS, e.g. `text/applyEdit`, `text/openFile`, etc. Many are not implemented, but can be inferred from the suggestion database. I was thinking of providing a separate server that would work as a facade for Enso LS, which would forward some messages, receive suggestion database updates from the Enso LS, and process some messages from LSP client by traversing the database.

I don't think serializing/deserializing JSON messages twice (or three times) is a sign of good architecture. The last thing I want to end up with is a facade JSON server sending messages to another JSON server (which is sending messages to 3rd server already).

All our code is written in Java and it is enough to have a single JSON serialization/deserialization - the one between VSCode & the JVM process. The rest should rather be handled via methods calls - they are faster and easier to debug.

**Comment by Akirathan** [comment]
@JaroslavTulach is the `CI: keep up to date` label here on purpose? Just today, there were 5 unnecessary sets of tests because of that.

File: tools/enso4igv/src/main/java/org/enso/tools/enso4igv/enso/EnsoStructure.java
Comment (Akirathan) [inline tools/enso4igv/src/main/java/org/enso/tools/enso4igv/enso/EnsoStructure.java]: A shortcut is fetching `BindingsMap` from the `moduleIr` and asking for `definedEntities`. Also, I assume that in the future we also want to be able to get imported symbols, right?

File: tools/enso4igv/package.json
Comment (JaroslavTulach) [inline tools/enso4igv/package.json:60]: Register support for Enso language via https://github.com/apache/netbeans/pull/6174 API that will be available in post 18.0.0 release of the [extension](https://marketplace.visualstudio.com/items?itemName=ASF.apache-netbeans-java).

**Comment by JaroslavTulach** [comment]
With the most recent changes in this PR and with most recent NetBeans and with symlink, so that `libenso_parser.so` can be found, I managed to show proper outline for an Enso file:

Enso Outline in VSCode

**Comment by JaroslavTulach** [comment]
If we want to play with this we need some build of Apache NetBeans VSCode 19.0.0. Until it is uploaded to Marketplace, an [RC3 build is available](https://builds.apache.org/job/Netbeans/job/netbeans-vscode/1786/) from Apache Jenkins.

**Comment by JaroslavTulach** [comment]
> ![Enso Outline in VSCode]...

New version of the extension

File: .github/workflows/enso4igv.yml
Comment (JaroslavTulach) [inline .github/workflows/enso4igv.yml]: structure:

>    lib: contains native JNI libraries and basic boot files. The naming policy of the libraries is in hands of the modules, as that it has worked well enough till now. If it proves unsufficient following conventions shall be obeyed: There are subdirectories under the lib directory that reflect the architecture (e.g. i386, obtained via System.getProperty ("os.arch")) and possibly also subdirectory identifying the system (e.g. i386/linux, obtained via System.getProperty ("os.name").toLowerCase ()). When the system is looking for a particular library it scans the deepest directories first and then the shallow ones (lib/i386/linux, lib/i386, lib). 


can you please post here what are the values of `os.name` and `os.arch`  when executed on Windows and (new) Mac?

File: .github/workflows/enso4igv.yml
Comment (JaroslavTulach) [inline .github/workflows/enso4igv.yml:42]: I assume we will need another mac build to create M1 binary.

File: .github/workflows/enso4igv.yml
Comment (radeusgd) [inline .github/workflows/enso4igv.yml]: Try
```suggestion
        env:
          RUSTFLAGS: "-C target-feature=-crt-static"
        run: |
          cargo build -p enso-parser-jni -Z unstable-options --out-dir target/lib/windows/amd64/
```

**Comment by JaroslavTulach** [comment]
Windows version is getting better, but on JDK-17 it fails (fixed by 87175ad) with:
```
SEVERE [org.openide.util.RequestProcessor]: Error in RequestProcessor org.netbeans.modules.java.lsp.server.protocol.TextDocumentServiceImpl$$Lambda$339/0x0000000801540690
java.lang.UnsupportedClassVersionError: org/enso/syntax2/Parser has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 61.0
	at java.base/java.lang.ClassLoader.defineClass1(Native Method)
	at java.base/java.lang.ClassLoader.defineClass(ClassLoader.java:1012)
	at org.netbeans.JarClassLoader.doLoadClass(JarClassLoader.java:314)
	at org.netbeans.ProxyClassLoader.selfLoadClass(ProxyClassLoader.java:260)
	at org.netbeans.ProxyClassLoader.doFindClass(ProxyClassLoader.java:189)
	at org.netbeans.ProxyClassLoader.loadClass(ProxyClassLoader.java:140)
	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:520)
	at org.enso.compiler.core.EnsoParser.<init>(EnsoParser.java:14)
	at org.enso.tools.enso4igv.enso.EnsoStructure.collectStructure(EnsoStructure.java:34)
Caused: java.lang.IllegalStateException
	at org.enso.tools.enso4igv.enso.EnsoStructure.collectStructure(EnsoStructure.java:42)
	at org.enso.tools.enso4igv.enso.EnsoStructure.getStructure(EnsoStructure.java:22)
```
with JDK 21 I am however seeing **the outline**:

outline in VSCode

**Comment by JaroslavTulach** [comment]
Screenshot on Intel based Mac:

image

**Comment by JaroslavTulach** [comment]
Please test [enso4vscode-1.36.106.vsix](https://github.com/enso-org/enso/actions/runs/9500540370) on Linux, Mac Intel, Mac Arm, and Windows. If the outline view isn't shown, then please run `jshell` and provide values of `System.getProperty("os.name")` and `System.getProperty("os.arch")`. Thank you.

**Comment by JaroslavTulach** [comment]
CCing @ola-lis, your help will be needed soon!

**Comment by radeusgd** [comment]
The extension is working ok for me on Windows.

image

How hard would it be to also display methods-defined-on-types in the outline? Most Enso methods are on a type, so that would be very helpful. But it is already an amazing step in a great direction 🎉

**Comment by radeusgd** [comment]
It doesn't work if I open a folder other than the `enso` repository root.

For example I have a folder where I keep some bug repros. It would be great if the tools could work there as well.

image

**Review by radeusgd — APPROVED** [review]

**Comment by Akirathan** [comment]
> How hard would it be to also display methods-defined-on-types in the outline? Most Enso methods are on a type, so that would be very helpful. But it is already an amazing step in a great direction

@radeusgd Shouldn't be hard. These instance methods are in IR. Once this PR is merged, I can try to provide this as a follow-up.

> Please test [enso4vscode-1.36.106.vsix](https://github.com/enso-org/enso/actions/runs/9500540370) on Linux, Mac Intel, Mac Arm, and Windows.Please test [enso4vscode-1.36.106.vsix](https://github.com/enso-org/enso/actions/runs/9500540370) on Linux, Mac Intel, Mac Arm, and Windows.Please test [enso4vscode-1.36.106.vsix](https://github.com/enso-org/enso/actions/runs/9500540370) on Linux, Mac Intel, Mac Arm, and Windows.


@JaroslavTulach The outline is displayed for me on Ubuntu 24.04, x64, but every time I click in an Enso source, I see NullPointerException on `textDocument/codeAction` message from the NetBeans LS extension:
image

**Comment by JaroslavTulach** [comment]
> > Please test [enso4vscode-1.36.106.vsix](https://github.com/enso-org/enso/actions/runs/9500540370) on Linux, Mac Intel, Mac Arm, and Windows.Please test [enso4vscode-1.36.106.vsix](https://github.com/enso-org/enso/actions/runs/9500540370) on Linux, Mac Intel, Mac Arm, and Windows.Please test [enso4vscode-1.36.106.vsix](https://github.com/enso-org/enso/actions/runs/9500540370) on Linux, Mac Intel, Mac Arm, and Windows.
> 
> @JaroslavTulach The outline is displayed for me on Ubuntu 24.04, x64, but every time I click in an Enso source, I see NullPointerException on `textDocument/codeAction` message from the NetBeans LS extension:

Great to hear it is working. This is the source of the NPE:
- https://github.com/apache/netbeans/pull/7466

Hopefully we can live with the NPE - otherwise we would have to wait for next release of NetBeans VSCode extension.

**Comment by JaroslavTulach** [comment]
> The extension is working ok for me on Windows.

Great!

> > How hard would it be to also display methods-defined-on-types in the outline? Most Enso methods are on a type, so that would be very helpful. But it is already an amazing step in a great direction
> 
> @radeusgd Shouldn't be hard. These instance methods are in IR. Once this PR is merged, I can try to provide this as a follow-up.

I happily leave the work up to you guys! I am glad the infrastructure seems to work (to some extent). The improvements in the outline shall be simple:

1. [write a test](https://github.com/enso-org/enso/pull/7054/files#diff-ba3c9b3f4096446cc7bb352a7976b0b9f161fcde7caa4b582e5d25a22cd97b17R37)
2. [fix it](https://github.com/enso-org/enso/pull/7054/files#diff-48e5e6c6cd09cd1d1398cb39973470e4d85a212d0df83d8240186235e55fb483R69)
3. [rebuild](https://github.com/enso-org/enso/tree/2ae1255aab7b22cec33797de29a12935a2ffe971/tools/enso4igv#building-vscode-extension) and test in VSCode
4. **repeat**!

You are both experienced `IR` users - your help with _outline_ & co. is welcomed!

File: lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java
Comment (JaroslavTulach) [inline lib/rust/parser/generate-java/java/org/enso/syntax2/Parser.java:12]: @kazcw, your review is needed (probably) because of this `Parser` change. Rather than trying to locate the library by `searchFromDirToTop` it, first of all just asks the system to load the library itself. The _NetBeans Runtime Container_ [specifies how to package the libraries](https://web.archive.org/web/20210118080511/https://platform.netbeans.org/articles/installation.html) for different architectures:

> lib: contains native JNI libraries and basic boot files. The naming policy of the libraries is in hands of the modules, as that it has worked well enough till now. If it proves unsufficient following conventions shall be obeyed: There are subdirectories under the lib directory that reflect the architecture (e.g. i386, obtained via System.getProperty ("os.arch")) and possibly also subdirectory identifying the system (e.g. i386/linux, obtained via System.getProperty ("os.name").toLowerCase ()). When the system is looking for a particular library it scans the deepest directories first and then the shallow ones (lib/i386/linux, lib/i386, lib). 

Currently the [enso4igv-1.36.106.nbm](https://github.com/enso-org/enso/actions/runs/9500540370) file contains:
```
$ unzip -l enso4igv-1.36.106.nbm | grep lib/ | cut -c31-
netbeans/modules/lib/
netbeans/modules/lib/aarch64/
netbeans/modules/lib/aarch64/libenso_parser.dylib
netbeans/modules/lib/enso_parser.dll
netbeans/modules/lib/libenso_parser.dylib
netbeans/modules/lib/libenso_parser.so
```
With some luck the `System.loadLibrary("enso_parser")` will load the right library on each of the supported systems.

**Review by kazcw — APPROVED** [review]
Parser change LGTM

File: .github/workflows/enso4igv.yml
Comment (JaroslavTulach) [reply .github/workflows/enso4igv.yml:42]: Done in 2ae1255aab7b22cec33797de29a12935a2ffe971

File: .github/workflows/enso4igv.yml
Comment (JaroslavTulach) [inline .github/workflows/enso4igv.yml:32]: One thing that might benefit from your review, @kazcw are these `cargo build` commands. Right now [they produce](https://github.com/enso-org/enso/actions/runs/9509862775?pr=7054) relatively huge artifacts:
parser artifacts
and that's the size after compression, the [real size of the generated files](https://github.com/enso-org/enso/actions/runs/9509862775/job/26213518760#step:4:5) is:
```
 4229656  modules/lib/aarch64/libenso_parser.dylib
 2207232  modules/lib/enso_parser.dll
 4035816  modules/lib/x86_64/libenso_parser.dylib
 8151600  modules/lib/libenso_parser.so
```
Any chance to make to make these files smaller?

**Comment by JaroslavTulach** [comment]
> It doesn't work if I open a folder other than the `enso` repository root.
> For example I have a folder where I keep some bug repros. It would be great if the tools could work there as well.

The version [v1.36.114](https://github.com/enso-org/enso/actions/runs/9509862775?pr=7054) does work for me for a single file on Mac:

Snímek obrazovky 2024-06-14 v 5 10 18

The outline is filled. Please note the _"Indexing completed."_ in the status bar - that's an indication the Apache NetBeans VSCode Extension has been activated. Also visible in _Output/Apache NetBeans Language Server_ output tab. Do you trust the folder?

main in windows

It also works in Windows including execution with Ctrl+F5.

**Review by Akirathan — APPROVED** [review]

**Review by 4e6 — APPROVED** [review]

**Review by AdRiley — APPROVED** [review]

**Review by hubertp — APPROVED** [review]

**Review by jdunkerley — APPROVED** [review]

**Review by Frizi — APPROVED** [review]

File: .github/workflows/enso4igv.yml
Comment (kazcw) [reply .github/workflows/enso4igv.yml:32]: `cargo build --release`

**Comment by JaroslavTulach** [comment]
@ola-lis, can you please release [enso4vscode-1.40.118.vsix](https://github.com/enso-org/enso/actions/runs/9516542525?pr=7054) on [VSCode marketplace](https://marketplace.visualstudio.com/items?itemName=Enso.enso4vscode)?

File: .github/workflows/enso4igv.yml
Comment (JaroslavTulach) [reply .github/workflows/enso4igv.yml:32]: Thank you. Trying in #10295

**Pull request merged**
